examples/nlp/language_modeling/megatron_gpt_mcore_batch_eval.py

```diff
@@ -0,0 +1,222 @@
+        ):
+            app_state.model_parallel_size = (
+                cfg.tensor_model_parallel_size
+                * cfg.pipeline_model_parallel_size
+                * cfg.get('expert_model_parallel_size', 1)
+            )
+            app_state.tensor_model_parallel_size = cfg.tensor_model_parallel_size
+            app_state.pipeline_model_parallel_size = cfg.pipeline_model_parallel_size
+            app_state.expert_model_parallel_size = cfg.get('expert_model_parallel_size', 1)
+            (
+                app_state.tensor_model_parallel_rank,
+                app_state.pipeline_model_parallel_rank,
+                app_state.expert_model_parallel_rank,
+                app_state.model_parallel_size,
+                app_state.data_parallel_size,
+                app_state.pipeline_model_parallel_split_rank,
+                app_state.virtual_pipeline_model_parallel_rank,
+            ) = fake_initialize_model_parallel(
+                world_size=app_state.model_parallel_size,
+                rank=trainer.global_rank,
+                tensor_model_parallel_size_=cfg.tensor_model_parallel_size,
+                pipeline_model_parallel_size_=cfg.pipeline_model_parallel_size,
+                pipeline_model_parallel_split_rank_=cfg.pipeline_model_parallel_split_rank,
+                expert_model_parallel_size_=cfg.get('expert_model_parallel_size', 1),
+            )
+        checkpoint_path = os.path.join(cfg.checkpoint_dir, cfg.checkpoint_name)
+        # checkpoint_path is a dir in case of distributed checkpointing
+        if not os.path.isdir(checkpoint_path):
+            # legacy checkpoint needs model parallel rank injection
+            checkpoint_path = inject_model_parallel_rank(os.path.join(cfg.checkpoint_dir, cfg.checkpoint_name))
+        model = MegatronGPTModel.load_from_checkpoint(checkpoint_path, hparams_file=cfg.hparams_file, trainer=trainer)
+    else:
+        raise ValueError("need at least a nemo file or checkpoint dir")
+
+    model.freeze()
+
+    # Have to turn off activations_checkpoint_method for inference
+    try:
+        model.model.language_model.encoder.activations_checkpoint_method = None
+    except AttributeError:
+        pass
+
+    args = Namespace
+    args.inference_batch_times_seq_len_threshold = cfg.inference_batch_times_seq_len_threshold
+    args.padded_vocab_size = model.padded_vocab_size
+    args.fp32_residual_connection = model.cfg.fp32_residual_connection
+    args.hidden_size = model.cfg.hidden_size
+    args.params_dtype = model.cfg.precision
+    args.max_batch_size = cfg.max_batch_size
+
+    # We need this wrapper since mcore generate uses tokenizer.detokenize, tokenizer.tokenize to encode and decode prompts
+    class MCoreTokenizerWrappper:
+        def __init__(self, tokenizer):
+            self.tokenizer = tokenizer
+            self.eod = tokenizer.eod
+            self.vocab_size = tokenizer.vocab_size
+
+        def detokenize(self, tokens):
```

Comment:
Can't we just add these methods to the `TokenizerSpec` in nemo so we don't need this wrapper-class?

Reply:
Hmm, this is very specific to mcore generate, so I didnt want to corrupt the original tokenizer spec class.

Reply:
We are also reworking tokenizers later maybe for mcore 0.9 or the next release, where nemo and mcore will use the same tokenizer. At that point will unify the apis